### PR TITLE
Simpler openapi gen - subprojects do themselves

### PR DESCRIPTION
**Pull request by thockin**
This involves moving the report files, but it allows me to delete the indirect variable and indirect array code in update-codgen.  As proud as I was of figuring that out, I am also ashamed of myself for doing it.

This is my atonement.

Depends on #117898

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh [hunk 1]
```diff
@@ -28,11 +28,17 @@ kube::codegen::gen_helpers \
...
+if [[ "${UPDATE_API_KNOWN_VIOLATIONS:-}" == "true" ]]; then
+    update_report="--update-report"
+fi
+
 kube::codegen::gen_openapi \
     --input-pkg-root k8s.io/apiextensions-apiserver/pkg \
     --extra-pkgs k8s.io/api/autoscaling/v1 `# needed for Scale type` \
     --output-pkg-root k8s.io/apiextensions-apiserver/pkg/generated \
     --output-base "$(dirname "${BASH_SOURCE[0]}")/../../.." \
+    --report-filename "${SCRIPT_ROOT}/api_violation_exceptions.list" \
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh:40]: I like the partitioning of the generation a lot

Not sure about locating the exceptions file at the root of the module, for a couple reasons:
* it's ugly (would be on the main page of https://github.com/kubernetes/apiextensions-apiserver, for example)
* the current location routes to API review due to https://github.com/kubernetes/kubernetes/tree/master/api, which we'd lose

is there an API package we could tuck it away into?

Comment (thockin) [reply staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh:40]: The primary requirement is that it has to run from the root of each module (standalone) and from k/k.

Options:

1) Put it in a subdir of each module
2) Put it in a dotfile of each module
3) pass `--report-filename` as an arg to each module's update-codegen.sh
4) use a hidden env var to pass the moral equivalent of `--report-filename` to each module
5) try to sniff up the directory tree and figure out if we are running from k/k or not

(1) and (2) are OK to me.  The rest all kind of stink.  We simulataneously want these repos to work as standalone things AND we want them to dump their report into a parent which might not exist.

If I am squeezed, I will suggest #3, but I'll have to hold me nose while I do it :)

We could do #1 with OWNERS ?

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh:40]: 1 seems ok and works with OWNERS

4 actually doesn't bother me as a tweak to the `UPDATE_API_KNOWN_VIOLATIONS` envvar we're already passing down to trigger this... is it worse to trigger off a non-empty `UPDATE_API_KNOWN_VIOLATIONS_FILE` envvar instead and require that to either be relative to $PWD or an absolute path?

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh:40]: > How about ...

any of those seem fine

> I don't really know what we do with these other than look at them and then approve anyway :)

We really do avoid growing them for new stuff, and it regularly catches name mismatches or failure to explicitly pick a list type

Comment (thockin) [reply staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh:40]: Or I can just use the old names

Comment (thockin) [reply staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh]: new commit pushed - take a look

File: hack/update-codegen.sh [hunk 2]
```diff
@@ -555,148 +515,73 @@ function codegen::openapi() {
...
-            vendor/k8s.io/kube-aggregator \
-            "${apimachinery_dirs[@]}"
-        )
-
-    # shellcheck disable=SC2034 # used indirectly
-    local APIEXTENSIONS_output_dir="staging/src/k8s.io/apiextensions-apiserver/pkg/generated/openapi"
-    # shellcheck disable=SC2034 # used indirectly
-    local APIEXTENSIONS_known_violations_file="api/api-rules/apiextensions_violation_exceptions.list"
-    # shellcheck disable=SC2034 # used indirectly
-    local APIEXTENSIONS_tag_files=()
```
Comment (Tusenka) [inline hack/update-codegen.sh:603]: why do we need remove this lines as well?

**Comment by jpbetz** [comment]
I did a full pass. This looks right to me.

It's great to have the subprojects handling their own generation. In addition to simplifying the main project, it makes the same projects better starting points.

**Comment by liggitt** [comment]
/lgtm

want to just merge this one and include https://github.com/kubernetes/kubernetes/pull/117898?

File: hack/update-codegen.sh (hunk 2 above)
Comment (thockin) [reply hack/update-codegen.sh:603]: All of the "indirect" logic is no longer needed - apiextensions, aggregator, examples, etc are handled inside the "subproject" logic.

**Comment by fedebongio** [comment]
Thank you Tim!
/triage accepted

**Pull request merged**